File: backend/app/schemas/crosswalk.py
"""Schemas for the Tier 3 crosswalk's atomic move-members endpoint (Path A, #328).

The move-members endpoint atomically updates BOTH the equivalence-group link
(`DatasetColumn.equivalence_group_id`) AND analysis-domain membership
(`AnalysisDomainMember`) in one transaction. This replaces today's "drag
changes EG link only, domain membership stays — phantom cells across
brackets" semantics with "drag = full move, both layers update together."

The endpoint also supports drag-to-Unassigned (`target_domain_id=None`,
`target_mode='strip'`), promoting two single-dataset members into a paired
EG (`target_mode='new_eg'`), and multi-select moves (N column_ids in one
transaction).

See #328 and the internal design notes
for the full architecture.
"""
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field, model_validator

from .analysis_domain import AnalysisDomainResponse


class MoveMembersRequest(BaseModel):
    """Request body for `POST /projects/{pid}/crosswalk/move-members`.

    `target_mode` is the discriminator:
      * `existing_eg` — assign all `column_ids` to the existing EG `target_eg_id`.
      * `new_eg` — create a new EG with `target_eg_label`, attach all `column_ids`.
      * `strip` — set `equivalence_group_id = None` for all `column_ids` (synthetic
        single-cell rows in the target domain, or fully unassigned).
    """
    column_ids: list[int] = Field(..., min_length=1)
    source_domain_id: int | None = None
    target_domain_id: int | None = None
    target_mode: Literal["existing_eg", "new_eg", "strip"]
    target_eg_id: int | None = None
    target_eg_label: str | None = None

    @model_validator(mode="after")
    def _validate_target_mode_payload(self):
        if self.target_mode == "existing_eg":
            if self.target_eg_id is None:
                raise ValueError(
                    "target_eg_id is required when target_mode='existing_eg'"
                )
            if self.target_eg_label is not None:
                raise ValueError(
                    "target_eg_label must be None when target_mode='existing_eg'"
                )
        elif self.target_mode == "new_eg":
            if not self.target_eg_label or not self.target_eg_label.strip():
                raise ValueError(
                    "target_eg_label is required when target_mode='new_eg'"
                )
            if self.target_eg_id is not None:
                raise ValueError(
                    "target_eg_id must be None when target_mode='new_eg'"
                )
        elif self.target_mode == "strip":
            if self.target_eg_id is not None or self.target_eg_label is not None:
                raise ValueError(
                    "target_eg_id and target_eg_label must both be None when target_mode='strip'"
                )

        if self.source_domain_id is None and self.target_domain_id is None:
            raise ValueError(
                "At least one of source_domain_id or target_domain_id must be set"
            )
        return self


class MoveMembersResponse(BaseModel):
    """Response from move-members.

    `source_domain` and `target_domain` are loaded fresh after the transaction
    commits. Either may be None if the corresponding side wasn't part of the
    move (e.g., promoting unassigned columns has no source domain).

    `dissolved_eg_ids` lists any equivalence groups that became empty as a
    result of the move and were auto-deleted in the same transaction.

    `recomputed_metric_ids` mirrors the swap endpoint's reporting for the
    domain_aggregate metrics that were synchronously recomputed.
    """
    source_domain: AnalysisDomainResponse | None = None
    target_domain: AnalysisDomainResponse | None = None
    dissolved_eg_ids: list[int] = Field(default_factory=list)
    recomputed_metric_ids: list[int] = Field(default_factory=list)

File: frontend/src/lib/api/crosswalk.ts
/**
 * Crosswalk API namespace — Path A's atomic move-members endpoint.
 *
 * `move-members` is the canonical primitive behind drag-to-bracket and
 * drag-to-unassigned (Tier 3 redesign #328). It atomically updates BOTH
 * the equivalence-group link AND analysis-domain membership in one
 * transaction, with post-mutation validators (#290 cross-dataset pairing,
 * #289 1:1-per-dataset) and source-EG auto-dissolve (#323).
 */
import api from './client'
import type { AnalysisDomainResponse } from './analysis-domains'

export type MoveMembersTargetMode = 'existing_eg' | 'new_eg' | 'strip'

export interface MoveMembersRequest {
  column_ids: number[]
  source_domain_id: number | null
  target_domain_id: number | null
  target_mode: MoveMembersTargetMode
  /** Required if target_mode === 'existing_eg' */
  target_eg_id?: number | null
  /** Required if target_mode === 'new_eg' */
  target_eg_label?: string | null
}

export interface MoveMembersResponse {
  source_domain: AnalysisDomainResponse | null
  target_domain: AnalysisDomainResponse | null
  dissolved_eg_ids: number[]
  recomputed_metric_ids: number[]
}

export const crosswalkApi = {
  moveMembers: (projectId: number, request: MoveMembersRequest) =>
    api
      .post<MoveMembersResponse>(
        `/projects/${projectId}/crosswalk/move-members`,
        request,
      )
      .then((res) => res.data),
}
